### ISSUE TITLE
[BRAINSTORMING] Virtual Gamepad Feature Talk and Issues

### ISSUE THREAD

**Issue by shinra358** [Ryochan7/DS4Windows#1312]
Added option to specify permanently connected output devices

And what is it used for?

**Comment by Yohoki** [comment]
Output Slots tab. It's a new tab that's been added between Auto Profiles and Settings.

**Comment by shinra358** [comment]
What's it's purpose? Is it to fake a controller to be always connected for pc games that don't have proper hotswapping support?

**Comment by Yohoki** [comment]
I'm looking at https://github.com/Ryochan7/DS4Windows/commit/8071fcfc28fdd23f9e9986595ccd53ac68415ace right now. Looks to me like it makes the virtual controller stay "plugged in" in the event that the physical controller comes. So if I'm reading it correctly, yes, it seem like it's a fix for games that don't support hotswapping correctly.

**Comment by shinra358** [comment]
We need to be able to use this on a per game bases though or all is for
nought. Does that already exist or am i overlooking it?

**Comment by mika-n** [comment]
#1163 topic is one scenario which is hopefully solved by this new persistent virtual controller feature.

**Comment by shinra358** [comment]
So are we right or wrong about the above hypothesis?

**Comment by Ryusennin** [comment]
"We need to be able to use this on a per game bases though or all is for nought."

I don't get your point. All games are potentially affected. Actually, I can easily force DS4W to switch ports without even launching a game. Making this feature 'per game' would be useless in that case.

**Comment by shinra358** [comment]
Hold on, please answer previous question to confirm that we're all on same
page.

**Comment by mika-n** [comment]
I still don't see what would be the use case to have this new "persistent virtual gamepad controller" feature as a profile specific option. It would nullify the whole purpose of the feature and what it tries to solve. Or some people would then complain why they have to enable this feature in all profiles to make it useful instead of having just one option.

Are you thinking of some multiplayer games in emulators where another play can drop in and out by connecting or disconnecting a gamepad (many console games behaves like this)?

**Comment by shinra358** [comment]
I have no idea what youre talking about bro. It's almost as if you dont
understand what i'm saying. Because i have no idea what you are talking
about. Because ive asked you to confirm or deny some things earlier but you
havent but in order to progress the convo, the thing that was asked must be
answered 😩.

So ill ask again in the simplest way i can. In your own words, what ate all
the things this feature achieves/does? And then we can go from there.

**Comment by mika-n** [comment]
I guess you haven't read the "ghost controller" issue I linked? It explains the problem in xinput device system and in ViGem driver.  The thing is that sometimes WinOS and/or ViGem driver leaves ghost devices behind when a virtual gamepad is disconnected. This causes problems in many games because when you re-connect the physical gamepad (without rebooting WinOS) it is assigned to a new xinput slot number. Many games see this as if you have now two gamepads or if you had mapped xinput slot#1 in game settings then now all in a sudden the gamepad is seen by the game as xinput slot#2 gamepad.

Also, there are often exclusive mode (=Hide DS4 Controller) issues in games if the physical gamepad was disconnected and re-connected while the game was running. Exclusive mode often fails after the next re-connection because the game had hicjacked the phystical gamepad device during the disconnect event. Most EA games (FIFA, NHL, etc) are notorious of this (lazy EA programmers).
https://github.com/Ryochan7/DS4Windows/wiki/Exclusive-Mode-(Hide-DS4-Controller-config-option)-tips-and-issues

Sometimes BT connection is accidentally disconnected or usb cable is loose and the gamepad is accidentally disconnected in the middle of gaming session. This persistent virtual output gamepad features helps to workaround issues of disconnection. If the game uses the virtual output controller as an input then the game won't see the disconnect event and this way it won't try to use new slot# or hijack the physical gamepad device if exclusive mode is used.

So, to me it looks like you haven't fully understood the problem of "xnput slot# rotation / ghost devices after disconnect - reconnection / hijacked physical gamepad device" issues. Please read links I provided and I'm sure you will see why this new feature is there. 

https://github.com/Ryochan7/DS4Windows/issues/877

https://github.com/Ryochan7/DS4Windows/issues/874

**Comment by shinra358** [comment]
That's exactly what i said in the beginning, that i didnt understand. I
gave guesses and i wanted you to tell me yes or no. You linked me to a
thread instead where it talked about several things as stated earlier. And
it seems like you said everything i just mentioned.

So that is what i was asking. It's not for what i thought it was for
earlier. So i guess i was looking for the confirmation of no. Seems purely
a slot issue fix and nothing more. So since it's now confirmed to be for
something else, then what i was asking after doesnt apply.

**Comment by shinra358** [comment]
haha, you know what..... I tried it myself and with it on, it still allows for disconnect prompts on normal games and at the same times, fixes connectivity issues for poorly ported games. so it does infact do what i guess but yes no per game thing is needed because everything works perfectly like I wanted without doing that. lol. For example, when in use, I can play DoA5 without having to restart the game if controller gets disconnected or if I didn't start with controller on and the game still pauses on Street Fighter IV if controller is disconnected. Great.


So, last questions:

1. If a controller is turned on, and you press [unplug], nothing happens. When [unplug] is pressed, can you trigger to turn the controllers off?

2. Can we keep the settings on ds4windows restart? As of now, we have to keep manually pressing [plug] each pc/ds4windows session. I want to be able to turn my pc on and go and not have to keep going inside the ds4windows when everything is already set up.

**Comment by Ryusennin** [comment]
Settings for permanent devices are saved in OutputSlots.xml file. If that file doesn't exist, you probably have a folder access issue. Hint: don't install DS4W in Program Files, use a custom directory.

**Comment by shinra358** [comment]
```
<?xml version="1.0" encoding="utf-8"?>
<!-- Made with DS4Windows version 2.1.2 -->

<OutputSlots app_version="2.1.2" />
```
That's the only thing in that file. And I have ds4windows located on a secondary drive (E)

**Comment by mika-n** [comment]
There should be something like this when a lot is assigned as "permanent". When you have set a slot as "permanent" and pressed "Accept" button then this xml file is not saved right away. When DS4Windows app is closed it will write out existing up-to-date config values to this and other xml files.

```
<OutputSlots app_version="2.1.2">
  <Slot idx="0">
    <DeviceType>X360</DeviceType>
  </Slot>
</OutputSlots>
```

**Comment by shinra358** [comment]
Maybe me having it on dynamic was the issue. Ill try permanent when i have
a break.

**Comment by shinra358** [comment]
Okay, that worked. Unfortunately, I ran into an issue where I cannot turn the controllers off and all of that tab greys out now both my controllers lights are stuck on even when ds4windows is stopped and off.

I turned on one controller only. it changed to p1 color. Then I turned it off. Then I turned on the other controller, lightbar was white.  So I pressed stopped on ds4 windows and started again. Turned one controller on again and 2 xbox controllers appeared in joy.cpl with one generic pad also listed. In which then I tried to turn the controller off but I couldn't.

Restarting ds4windows at this point turned the lights off and back on properly. But I am still unable to turn controllers back off via psbutton + start.

battery level for both are 50 something and 80 something. everything works as it should i guess on dynamic. But not on permanent.
